osu.Game/Rulesets/Objects/Drawables/DrawableHitObject.cs

```diff
@@ -336,23 +336,32 @@ protected sealed override void SkinChanged(ISkinSource skin, bool allowFallback)
         {
             base.SkinChanged(skin, allowFallback);
 
-            updateAccentColour();
+            updateComboColour();
 
             ApplySkin(skin, allowFallback);
 
             if (IsLoaded)
                 updateState(State.Value, true);
         }
 
-        private void updateAccentColour()
+        private void updateComboColour()
         {
             if (HitObject is IHasComboInformation combo)
             {
                 var comboColours = CurrentSkin.GetConfig<GlobalSkinColours, IReadOnlyList<Color4>>(GlobalSkinColours.ComboColours)?.Value;
-                AccentColour.Value = comboColours?.Count > 0 ? comboColours[combo.ComboIndex % comboColours.Count] : Color4.White;
+                UpdateComboColour(comboColours?.Count > 0 ? comboColours[combo.ComboIndex % comboColours.Count] : Color4.White, comboColours);
             }
         }
 
+        /// <summary>
+        /// Called when a combo colour change is proposed.
+        /// </summary>
+        /// <param name="proposedColour">The proposed combo colour, based off the combo index.</param>
+        /// <param name="comboColours">A list of combo colours provided by the beatmap or skin. Can be null if not available.</param>
+        protected virtual void UpdateComboColour(Color4 proposedColour, IReadOnlyList<Color4> comboColours)
+        {
+        }
```

Comment:
It's probably because I don't fully know where you plan taking this forward, but I'm not feeling this API too much. I'd probably prefer a pure function myself (return `Color4` and then assign it where it needs to go).

I understand *why* the default implementation is empty (if it fell back to the old behaviour, there'd be no way of undoing it), and I'm guessing it's done like this to allow ruleset creators to change more than just one accent colour...?

Reply:
we could feasibly leave the default implementation in there and cancel it out with no-base-call, although not sure how i feel about that since we try to avoid relying on base calls for externally exposed APIs.

Reply:
The proposal is to make this return a Color4.

```csharp
DHO:
    protected virtual Color4 GetComboColour(IReadOnlyList<Color4> comboColours) => comboColours?.Count > 0 ? comboColours[combo.ComboIndex % comboColours.Count] : Color4.White;

DCHO:
    protected override Color4 GetComboColour(IReadOnlyList<Color4> comboColours) => comboColour?.Count > 0 ? comboColours[hitObject.IndexInBeatmap % comboColours.Count] : Color4.White;
```

Though as I mentioned yesterday, the whole `AccentColour` stuff feels super shocking to me. It's publicly exposed and because of that it loses all meaning in the context of an `IHasComboInformation` object - whatever accent colour you set on a `DHO` may or may not get replaced by the combo colours based on certain variables including timing.

So based on that it is my opinion that this shouldn't be a pure method unless `AccentColour` itself was only privately mutable, and that has its own issues to resolve too such as how do you set the accent colour of mania hitobjects, what should the value of the exposed colour be if there is no combo, or should it even _exist_ inside `DHO` (which again, I mentioned yesterday).

Reply:
I'm also fine with moving `AccentColour` into each ruleset if we decide that's the correct direction.

Reply:
Good point, the fact of `AccentColour` being public makes my proposal entirely unviable.

Separating per ruleset would be probably pretty good, as it eliminates the problem of trying to come up with a good abstraction for everything. I'm not even sure how a generic solution for all rulesets could look like and I'd probably have to sit down for a couple of hours to see if one is even viable.

Not sure it's something that needs to be done right now, though. Could roll with this as a solution for now until it's no longer tenable.

Reply:
Let's go with that then, make it a pure method and then we can consider moving it out into a `DrawableHitObjectWithComboColour` at a later point. Important part is getting the API into a steady state, because anything further from there won't be too hard to change.